File: sirius-core/src/main/java/org/starcoin/util/EncryptUtil.java
package org.starcoin.util;

import org.bouncycastle.jce.provider.BouncyCastleProvider;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.security.*;

/**
 * Download Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files see
 * https://stackoverflow.com/questions/6481627/java-security-illegal-key-size-or-default-parameters
 */
public class EncryptUtil {

    static {
        Security.addProvider(new BouncyCastleProvider());
    }

    public static byte[] encrypt(PublicKey publicKey, byte[] data) {
        try {
            // 参阅:
            // 1. http://www.flexiprovider.de/examples/ExampleECIES.html
            // 2. https://crypto.stackexchange.com/questions/12823/ecdsa-vs-ecies-vs-ecdh
            // 3. https://crypto.stackexchange.com/questions/24516/ecdsa-for-encryption

            Cipher cipher = Cipher.getInstance("ECIES", "BC");
            cipher.init(Cipher.ENCRYPT_MODE, publicKey);
            return cipher.doFinal(data);
        } catch (NoSuchAlgorithmException
                | InvalidKeyException
                | NoSuchPaddingException
                | BadPaddingException
                | NoSuchProviderException
                | IllegalBlockSizeException e) {
            throw new RuntimeException(e);
        }
    }

    public static byte[] decrypt(PrivateKey privateKey, byte[] encrypted) {
        try {
            Cipher cipher = Cipher.getInstance("ECIES", "BC");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);

            return cipher.doFinal(encrypted);
        } catch (NoSuchProviderException
                | NoSuchAlgorithmException
                | InvalidKeyException
                | NoSuchPaddingException
                | BadPaddingException
                | IllegalBlockSizeException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/util/EncryptUtil.kt
package org.starcoin.sirius.util

import org.bouncycastle.jce.provider.BouncyCastleProvider
import java.security.PrivateKey
import java.security.PublicKey
import java.security.Security
import javax.crypto.Cipher

/**
 * Download Java Cryptography Extension (JCE) Unlimited Strength Jurisdiction Policy Files see
 * https://stackoverflow.com/questions/6481627/java-security-illegal-key-size-or-default-parameters
 */
object EncryptUtil {

    init {
        Security.addProvider(BouncyCastleProvider())
    }

    fun encrypt(publicKey: PublicKey, data: ByteArray): ByteArray {
        // 参阅:
        // 1. http://www.flexiprovider.de/examples/ExampleECIES.html
        // 2. https://crypto.stackexchange.com/questions/12823/ecdsa-vs-ecies-vs-ecdh
        // 3. https://crypto.stackexchange.com/questions/24516/ecdsa-for-encryption

        val cipher = Cipher.getInstance("ECIES", "BC")
        cipher.init(Cipher.ENCRYPT_MODE, publicKey)
        return cipher.doFinal(data)
    }

    fun decrypt(privateKey: PrivateKey, encrypted: ByteArray): ByteArray {
        val cipher = Cipher.getInstance("ECIES", "BC")
        cipher.init(Cipher.DECRYPT_MODE, privateKey)

        return cipher.doFinal(encrypted)

    }
}
